File: DFS_BFS/127.word_ladder.py
# This solution TLE

# from collections import deque as Queue

# class Solution:
#     def helper(self, x, y):
#         if len(x) == len(y):
#             diff = 0
#             i = 0
#             while i < len(x):
#                 if x[i] != y[i]:
#                     diff += 1
#                 i += 1
#             return diff == 1
#         else:
#             return False
    
#     def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
#         adjacent = {}
#         N = len(wordList)
#         for x in wordList:
#             adjacent[x] = []
#             for y in wordList:
#                 if self.helper(x, y):
#                     adjacent[x].append(y)
#         if endWord not in adjacent:
#             return 0
#         else:
#             visited, distance = {}, {}
#             for word in wordList:
#                 visited[word] = 0
#                 distance[word] = -1
#             visited[endWord] = 1
#             distance[endWord] = 0
#             queue = Queue()
#             queue.append(endWord)
#             while len(queue) != 0:
#                 u = queue.popleft()
#                 for v in adjacent[u]:
#                     if visited[v] == 0:
#                         visited[v] = 1
#                         distance[v] = distance[u] + 1
#                         queue.append(v)
#             ans = 3 * N
#             for u in distance:
#                 if distance[u] != -1:
#                     if self.helper(u, beginWord):
#                         ans = min(ans, distance[u]+2)
#                     elif u == beginWord:
#                         ans = min(ans, distance[u]+1)
#             return ans if ans <= N+1 else 0

from collections import deque as Queue, defaultdict

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        transform = defaultdict(list)
        for word in wordList:
            for i in range(len(word)):
                transform[word[:i]+'*'+word[i+1:]].append(word)
        visited = {}
        for word in wordList:
            visited[word] = 0
        visited[beginWord] = 1
        queue = Queue()
        queue.append((beginWord, 1))
        while len(queue) != 0:
            word, level = queue.popleft()
            for i in range(len(word)):
                if word[:i]+'*'+word[i+1:] in transform:
                    for x in transform[word[:i]+'*'+word[i+1:]]:
                        if x == endWord:
                            return level+1
                        if visited[x] == 0:
                            visited[x] = 1
                            queue.append((x, level+1))
        return 0

File: C++/0.Comments/Breadth-First Seach/Word Ladder Series/127. Word Ladder.cpp
#include <unordered_map>
#include <queue>
#include <unordered_set>

class Solution {
public:
    
    vector<string> pattern(string& word) {
        string s = word;
        vector<string> res;
        for (int i = 0; i < s.size(); i++) {
            char temp = s[i];
            s[i] = '*';
            res.push_back(s);
            s[i] = temp;
        }
        return res;
    }
    
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        // 构造邻接字典
        unordered_map<string, vector<string>> neighbors;
        unordered_map<string, vector<string>> patterns;
        for (string& w : wordList) {
            patterns[w] = pattern(w);
            for (string& p : patterns[w]) {
                neighbors[p].push_back(w);
            }
        }
        // endword不在字典里
        if (patterns.find(endWord) == patterns.end()) return 0;
        // beginWord不在字典里
        if (patterns.find(beginWord) == patterns.end()) {
            patterns[beginWord] = pattern(beginWord);
        }
        // 构造目标匹配
        unordered_set<string> target;
        for (string& p : patterns[endWord]) target.insert(p);
        // 广度优先搜索
        queue<string> que;
        que.push(beginWord);
        unordered_set<string> visited;
        visited.insert(beginWord);
        int level = 1;
        while (!que.empty()) {
            int len = que.size();
            for (int i = 0; i < len; i++) {
                string front = que.front();
                for (string& p : patterns[front]) {
                    if (target.find(p) != target.end()) {
                        return level+1;
                    }
                    else {
                        for (string& n : neighbors[p]) {
                            if (visited.find(n) == visited.end()) {
                                que.push(n);
                                visited.insert(n);
                            }
                        }
                    }
                }
                que.pop();
            }
            level++;
        }
        return 0;
    }
};